src/command/subgraph/publish.rs

```diff
@@ -74,4 +126,84 @@ impl Publish {
             publish_response,
         })
     }
+
+    // output stream is optionally injected (defaulting to stdout) so we can test this function
+    pub fn handle_invalid_routing_url(
+        maybe_prompt_response: Option<bool>,
+        maybe_output: Option<&mut dyn std::io::Write>,
+    ) -> Option<RoverOutput> {
+        let default_output = &mut std::io::stdout();
+        let output = maybe_output.unwrap_or(default_output);
+        if let Some(prompt_response) = maybe_prompt_response {
+            if prompt_response {
+                None
+            } else {
+                writeln!(output, "Publish cancelled by user")
+                    .expect("Could not write to provided output stream");
+                Some(RoverOutput::EmptySuccess)
+            }
+        } else {
+            // if we're not in a tty, we didn't prompt. let's print a warning but
+            // publish anyway.
+            writeln!(
+                output,
+                "{} In a future major version of Rover, the `--allow-invalid-routing-url` flag will be required to publish a subgraph with an invalid routing URL in CI.",
+                Style::WarningPrefix.paint("WARN:")
+            ).expect("Could not write to provided output stream");
+            writeln!(
+                output,
+                "{} Found an invalid URL, but we can't prompt in a non-interactive environment. Publishing anyway.",
+                Style::WarningPrefix.paint("WARN:")
+            ).expect("Could not write to provided output stream");
+            None
+        }
+    }
+
+    pub fn prompt_for_publish() -> Option<bool> {
+        if !atty::is(atty::Stream::Stdout) {
+            // q for Avery: not actually sure why I can't implicitly return
+            // `None` here
+            return None;
+        }
+
+        match prompt_confirm_default_no(
+            "Found an invalid URL, would you still like to publish? [y/N]: ",
+        ) {
+            Ok(response) => Some(response),
+            _ => panic!("Expected a response in TTY environment"),
+        }
```

Comment:
i'd pass the user input to this function so we can use it in the prompt

```suggestion
        if prompt_confirm_default_no(
            format!("{invalid_routing_url} is not a valid routing URL. Continuing the publish will make this subgraph unreachable by your supergraph. Would you still like to publish? [y/N]: ",
        )? {
            Err(anyhow!("Your publish has been cancelled, no changes have been made to your supergraph."))
        } else {
          Ok(())
        }
```

not sure if this will compile directly or not but I think returning the error here is a bit clearer than handling it in the caller